zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/LazyHttpClientImpl.java

```diff
@@ -60,32 +59,10 @@
   EndpointGroup getEndpoint() {
     EndpointGroup initial = initialEndpoints.get();
     // Only health-check when there are alternative endpoints. There aren't when instanceof Endpoint
-    if (initial instanceof Endpoint) return initial;
+    if (initial instanceof Endpoint || !healthCheck.isEnabled()) return initial;
 
     // Wrap the result when health checking is enabled.
-    EndpointGroup result = initial;
```

Comment:
I might have misunderstood https://github.com/line/armeria/pull/2837#issuecomment-650105705 by @trustin or maybe the two tests below mentioning TODO are drifting for a different reason

Reply:
with or without this change, the health check code doesn't work anymore. for example, if I revert it, one of the tests that checks that a server can blip off -> on .. that one fails on the on part.

Reply:
This change itself looks like what I'd expect after the async endpoint selection change. Perhaps a timing issue that needs to tweak the timeouts in the test with this change added, or possibly a bug introduced? /cc @trustin

Reply:
Would you try to increase the connect timeout of the client factory?

```java
ClientFactory.builder()
             .channelOption(ChannelOption.CONNECT_TIMEOUT_MILLIS, ...)
             ...
             .build()
```

Reply:
we currently set like this..

```java
    // Elasticsearch 7 never returns a response when receiving an HTTP/2 preface instead of the more
    // valid behavior of returning a bad request response, so we can't use the preface.
    // TODO: find or raise a bug with Elastic
    return builder.useHttp2Preface(false)
      .connectTimeoutMillis(es.getTimeout())
      .meterRegistry(meterRegistry)
      .build();
```

The unit test value for timeout is 200ms

The following test fails

```java
  @Test public void notHealthyThenHealthyThenNotHealthy() {
    server1Health.setHealthy(false);
    server2Health.setHealthy(false);

    try (ElasticsearchStorage storage = context.getBean(ElasticsearchStorage.class)) {
      CheckResult result = storage.check();
      assertThat(result.ok()).isFalse();

      server2Health.setHealthy(true);

      // Health check interval is 100ms
      await().timeout(300, TimeUnit.MILLISECONDS).untilAsserted(() ->
        assertThat(storage.check().ok()).isTrue());

      server2Health.setHealthy(false);

      // Health check interval is 100ms
      await().timeout(300, TimeUnit.MILLISECONDS).untilAsserted(() ->
        assertThat(storage.check().ok()).isFalse()); <<< this never becomes false!
    }
  }
```

Reply:
This is the other that fails
```java
  // TODO: FIXME
  @Test public void healthyThenNotHealthyThenHealthy() {
    try (ElasticsearchStorage storage = context.getBean(ElasticsearchStorage.class)) {
      CheckResult result = storage.check();
      assertThat(result.ok()).isTrue();

      server1Health.setHealthy(false);
      server2Health.setHealthy(false);

      // Health check interval is 100ms
      await().timeout(300, TimeUnit.MILLISECONDS).untilAsserted(() ->
        assertThat(storage.check().ok()).isFalse()); <<< this never becomes false

      server1Health.setHealthy(true);

      // Health check interval is 100ms
      await().timeout(300, TimeUnit.MILLISECONDS).untilAsserted(() ->
        assertThat(storage.check().ok()).isTrue());
    }
  }
```

Reply:
there could be a red herring also.. if for example the mocks act differently than before.. running ITElasticsearchHealthCheck on this branch should fail on those two tests

```java
  static final SettableHealthChecker server1Health = new SettableHealthChecker(true);

  @ClassRule public static ServerRule server1 = new ServerRule() {
    @Override protected void configure(ServerBuilder sb) {
      sb.service("/", (ctx, req) -> VERSION_RESPONSE.toHttpResponse());
      sb.service("/_cluster/health", HealthCheckService.of(server1Health));
      sb.serviceUnder("/_cluster/health/", (ctx, req) -> GREEN_RESPONSE.toHttpResponse());
    }
  };

  static final SettableHealthChecker server2Health = new SettableHealthChecker(true);

  @ClassRule public static ServerRule server2 = new ServerRule() {
    @Override protected void configure(ServerBuilder sb) {
      sb.service("/", (ctx, req) -> VERSION_RESPONSE.toHttpResponse());
      sb.service("/_cluster/health", HealthCheckService.of(server2Health));
      sb.serviceUnder("/_cluster/health/", (ctx, req) -> GREEN_RESPONSE.toHttpResponse());
    }
  };
```